src/validate.js

```diff
@@ -1,169 +1,172 @@
+        const jsonName = name.replace(/\.ya?ml$/, '.json');
+        this.rawSchemas[jsonName] = json;
 
-      if (schema.id) {
-        debug('Schema incorrectly attempts to set own id: %s', name);
-        throw new Error('Schema ' + path.join(root, name) + ' attempts to set own id!');
-      }
-      let jsonName = name.replace(/\.ya?ml$/, '.json');
-      schema.id = libUrls.schema(cfg.rootUrl, cfg.serviceName, jsonName + '#');
+        const schema = render(json, TASKCLUSTER_SCHEMA_SCHEME, this.cfg.serviceName, this.cfg.constants);
 
-      ajv.addSchema(schema);
-      debug('Loaded schema with id of "%s"', schema.id);
-      let content = JSON.stringify(schema, undefined, 4);
-      if (!content) {
-        throw new Error('Schema %s has invalid content!', name);
-      }
-      schemas[jsonName] = content;
-    } catch (err) {
-      // walk swallows errors, so we must raise them ourselves
-      if (!walkErr) {
-        walkErr = err;
+        if (schema.id) {
+          debug('Schema incorrectly attempts to set own id: %s', name);
+          throw new Error('Schema ' + path.join(root, name) + ' attempts to set own id!');
+        }
+
+        // We use a `taskcluster:` scheme until rootUrl is known later
+        schema.id = libUrls.schema(TASKCLUSTER_SCHEMA_SCHEME, this.cfg.serviceName, jsonName + '#');
+        this.schemas[jsonName] = schema;
+      } catch (err) {
+        // walk swallows errors, so we must raise them ourselves
+        if (!walkErr) {
+          walkErr = err;
+        }
       }
+    }}});
+    if (walkErr) {
+      throw walkErr;
     }
-  }}});
-  if (walkErr) {
-    throw walkErr;
-  }
-  debug('finished walking tree of schemas');
-
-  if (cfg.publish) {
-    debug('Publishing schemas');
-    assert(cfg.aws, 'Can\'t publish without aws credentials.');
-    let s3Provider = cfg.s3Provider;
-    if (!s3Provider) {
-      debug('Using default s3 client');
-      s3Provider = new aws.S3(cfg.aws);
-    }
-    await Promise.all(_.map(schemas, (content, name) => {
-      return publish.s3(
-        s3Provider,
-        cfg.bucket,
-        `${cfg.serviceName}/`,
-        name,
-        content
-      );
-    }));
+    debug('finished walking tree of schemas');
   }
 
-  if (cfg.writeFile) {
-    debug('Writing schema to local file');
-    let dir = 'rendered_schemas';
-    rimraf.sync(dir);
-    _.forEach(schemas, (content, name) => {
-      const file = path.join(dir, name);
-      const subdir = path.dirname(file);
-      mkdirp.sync(subdir);
-      publish.writeFile(file, content);
+  async validator(rootUrl) {
+    const ajv = Ajv({useDefaults: true, format: 'full', verbose: true, allErrors: true});
+    this.schemas = _.mapValues(this.rawSchemas, (schema, name) => {
+      // Re-render to update refs to include rootUrl
+      schema.id = libUrls.schema(rootUrl, this.cfg.serviceName, name + '#');
+      schema = render(schema, rootUrl, this.cfg.serviceName, this.cfg.constants);
```

Comment:
I think these need to go in the opposite order :)

Reply:
Also, it looks like this modifies SchemaSet.schemas in-place.  That's kind of icky especially from a testing perspective.  Instead, could the returned validator have a `.schemas` property of its own that is rooted?  Alternatively, maybe a schema set has a `rootedSchemas(rootUrl)` method?